# Day 13 - Overfitting and Regularization

## ABTalks 60-Day Data Science Challenge

Comparing Linear Regression, Ridge Regression, and Lasso Regression to understand overfitting, regularization, and model generalization.

In [ ]:
import pandas as pd

df = pd.read_csv("superstore_sales_cleaned.csv")

df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

df['Shipping Days'] = (
    df['Ship Date'] - df['Order Date']
).dt.days

df['Order Year'] = df['Order Date'].dt.year

df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Shipping Days,Order Year
0,1,CA-2017-152156,2017-11-08,2017-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,3,2017
1,2,CA-2017-152156,2017-11-08,2017-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,2017
2,3,CA-2017-138688,2017-06-12,2017-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,4,2017
3,4,US-2016-108966,2016-10-11,2016-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,7,2016
4,5,US-2016-108966,2016-10-11,2016-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,7,2016


In [ ]:
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

df['Shipping Days'] = (
    df['Ship Date'] - df['Order Date']
).dt.days

df['Order Year'] = df['Order Date'].dt.year

df.head()

In [2]:
features = [
    'Shipping Days',
    'Order Year',
    'Category',
    'Sub-Category',
    'Region',
    'Ship Mode',
    'Segment'
]

X = df[features]
y = df['Sales']

print("Input features:")
print(X.columns.tolist())

print("\nTarget:")
print(y.name)

Input features:
['Shipping Days', 'Order Year', 'Category', 'Sub-Category', 'Region', 'Ship Mode', 'Segment']

Target:
Sales


In [3]:
X_encoded = pd.get_dummies(
    X,
    columns=[
        'Category',
        'Sub-Category',
        'Region',
        'Ship Mode',
        'Segment'
    ],
    dtype=int
)

print("Before encoding:", X.shape)
print("After encoding:", X_encoded.shape)

Before encoding: (9800, 7)
After encoding: (9800, 33)


In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_encoded,
    y,
    test_size=0.2,
    random_state=42
)

print("Training data:", X_train.shape)
print("Testing data:", X_test.shape)

Training data: (7840, 33)
Testing data: (1960, 33)


In [5]:
from sklearn.linear_model import LinearRegression

baseline_model = LinearRegression()

baseline_model.fit(X_train, y_train)

print("Baseline Linear Regression trained successfully!")

Baseline Linear Regression trained successfully!


In [6]:
from sklearn.metrics import mean_absolute_error, r2_score

# Predictions
baseline_train_pred = baseline_model.predict(X_train)
baseline_test_pred = baseline_model.predict(X_test)

# MAE
baseline_train_mae = mean_absolute_error(y_train, baseline_train_pred)
baseline_test_mae = mean_absolute_error(y_test, baseline_test_pred)

# R²
baseline_train_r2 = r2_score(y_train, baseline_train_pred)
baseline_test_r2 = r2_score(y_test, baseline_test_pred)

print("Baseline Model Performance")
print("---------------------------")
print("Train MAE:", baseline_train_mae)
print("Test MAE:", baseline_test_mae)
print("Train R²:", baseline_train_r2)
print("Test R²:", baseline_test_r2)

Baseline Model Performance
---------------------------
Train MAE: 198.79142447402415
Test MAE: 238.76911115425153
Train R²: 0.20481291573556326
Test R²: 0.17859549667597163


In [7]:
from sklearn.linear_model import Ridge

ridge_model = Ridge(alpha=1.0)

ridge_model.fit(X_train, y_train)

print("Ridge Regression trained successfully!")

Ridge Regression trained successfully!


In [8]:
ridge_train_pred = ridge_model.predict(X_train)
ridge_test_pred = ridge_model.predict(X_test)

ridge_train_mae = mean_absolute_error(y_train, ridge_train_pred)
ridge_test_mae = mean_absolute_error(y_test, ridge_test_pred)

ridge_train_r2 = r2_score(y_train, ridge_train_pred)
ridge_test_r2 = r2_score(y_test, ridge_test_pred)

print("Ridge Regression Performance")
print("----------------------------")
print("Train MAE:", ridge_train_mae)
print("Test MAE:", ridge_test_mae)
print("Train R²:", ridge_train_r2)
print("Test R²:", ridge_test_r2)

Ridge Regression Performance
----------------------------
Train MAE: 198.76521324827976
Test MAE: 238.78471795206008
Train R²: 0.204798188223004
Test R²: 0.17841055432177944


In [9]:
from sklearn.linear_model import Lasso

lasso_model = Lasso(alpha=1.0)

lasso_model.fit(X_train, y_train)

print("Lasso Regression trained successfully!")

Lasso Regression trained successfully!


In [10]:
lasso_train_pred = lasso_model.predict(X_train)
lasso_test_pred = lasso_model.predict(X_test)

lasso_train_mae = mean_absolute_error(y_train, lasso_train_pred)
lasso_test_mae = mean_absolute_error(y_test, lasso_test_pred)

lasso_train_r2 = r2_score(y_train, lasso_train_pred)
lasso_test_r2 = r2_score(y_test, lasso_test_pred)

print("Lasso Regression Performance")
print("----------------------------")
print("Train MAE:", lasso_train_mae)
print("Test MAE:", lasso_test_mae)
print("Train R²:", lasso_train_r2)
print("Test R²:", lasso_test_r2)

Lasso Regression Performance
----------------------------
Train MAE: 199.1710997124638
Test MAE: 239.68265489688574
Train R²: 0.20302093915807629
Test R²: 0.1716551768627953


In [11]:
comparison = pd.DataFrame({
    'Model': ['Linear Regression', 'Ridge', 'Lasso'],
    'Train MAE': [
        baseline_train_mae,
        ridge_train_mae,
        lasso_train_mae
    ],
    'Test MAE': [
        baseline_test_mae,
        ridge_test_mae,
        lasso_test_mae
    ],
    'Train R²': [
        baseline_train_r2,
        ridge_train_r2,
        lasso_train_r2
    ],
    'Test R²': [
        baseline_test_r2,
        ridge_test_r2,
        lasso_test_r2
    ]
})

comparison

,Model,Train MAE,Test MAE,Train R²,Test R²
0,Linear Regression,198.791424,238.769111,0.204813,0.178595
1,Ridge,198.765213,238.784718,0.204798,0.178411
2,Lasso,199.171100,239.682655,0.203021,0.171655


In [12]:
comparison['R² Gap'] = (
    comparison['Train R²'] - comparison['Test R²']
)

comparison

,Model,Train MAE,Test MAE,Train R²,Test R²,R² Gap
0,Linear Regression,198.791424,238.769111,0.204813,0.178595,0.026217
1,Ridge,198.765213,238.784718,0.204798,0.178411,0.026388
2,Lasso,199.171100,239.682655,0.203021,0.171655,0.031366


In [13]:
import numpy as np

lasso_coefficients = pd.Series(
    lasso_model.coef_,
    index=X_train.columns
)

near_zero = np.sum(np.isclose(lasso_coefficients, 0))

print("Total coefficients:", len(lasso_coefficients))
print("Coefficients near zero:", near_zero)

Total coefficients: 33
Coefficients near zero: 14


In [14]:
zero_features = lasso_coefficients[
    np.isclose(lasso_coefficients, 0)
]

print("Features with coefficients near zero:")
print(zero_features)

Features with coefficients near zero:
Order Year                  0.0
Category_Technology         0.0
Sub-Category_Binders        0.0
Sub-Category_Bookcases      0.0
Sub-Category_Chairs         0.0
Sub-Category_Phones        -0.0
Region_South                0.0
Region_West                -0.0
Ship Mode_First Class      -0.0
Ship Mode_Same Day         -0.0
Ship Mode_Second Class      0.0
Ship Mode_Standard Class   -0.0
Segment_Consumer           -0.0
Segment_Corporate          -0.0
dtype: float64


In [15]:
best_mae_model = comparison.loc[
    comparison['Test MAE'].idxmin(), 'Model'
]

best_r2_model = comparison.loc[
    comparison['Test R²'].idxmax(), 'Model'
]

print("Best model by Test MAE:", best_mae_model)
print("Best model by Test R²:", best_r2_model)

Best model by Test MAE: Linear Regression
Best model by Test R²: Linear Regression


## Day 13 Conclusion

### Model Comparison

Three regression models were compared:

- Linear Regression
- Ridge Regression
- Lasso Regression

Linear Regression achieved the best test performance with:
- Test MAE: 238.77
- Test R²: 0.179

Ridge Regression produced almost identical results to Linear Regression, showing that regularization with alpha=1.0 had only a small effect.

Lasso Regression reduced 14 out of 33 coefficients to approximately zero, demonstrating feature selection. However, its test performance was slightly lower.

### Overfitting Analysis

The train-test R² gaps were small:

- Linear Regression: 0.026
- Ridge: 0.026
- Lasso: 0.031

Therefore, there was no strong evidence of overfitting in this experiment.

### Key Learning

High training performance alone is not enough. Models should also perform well on unseen data.

Regularization can control model complexity and reduce the influence of unnecessary features, but it does not always improve test performance. The effect depends on the dataset and model.

In [1]:
# Remove the most important feature: Sub-Category
features_without_subcategory = [
    'Shipping Days',
    'Order Year',
    'Category',
    'Region',
    'Ship Mode',
    'Segment'
]

X_reduced = df[features_without_subcategory]

X_reduced_encoded = pd.get_dummies(
    X_reduced,
    columns=[
        'Category',
        'Region',
        'Ship Mode',
        'Segment'
    ],
    dtype=int
)

print("Original features:", X_encoded.shape[1])
print("Features after removal:", X_reduced_encoded.shape[1])

NameError: name 'df' is not defined

In [2]:
import pandas as pd

df = pd.read_csv("superstore_sales_cleaned.csv")

df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

df['Shipping Days'] = (
    df['Ship Date'] - df['Order Date']
).dt.days

df['Order Year'] = df['Order Date'].dt.year

print(df.shape)

(9800, 20)


In [3]:
features_without_subcategory = [
    'Shipping Days',
    'Order Year',
    'Category',
    'Region',
    'Ship Mode',
    'Segment'
]

X_reduced = df[features_without_subcategory]

X_reduced_encoded = pd.get_dummies(
    X_reduced,
    columns=[
        'Category',
        'Region',
        'Ship Mode',
        'Segment'
    ],
    dtype=int
)

print("Features after removal:", X_reduced_encoded.shape[1])

Features after removal: 16


In [4]:
from sklearn.model_selection import train_test_split

y = df['Sales']

X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced = train_test_split(
    X_reduced_encoded,
    y,
    test_size=0.2,
    random_state=42
)

print("Training data:", X_train_reduced.shape)
print("Testing data:", X_test_reduced.shape)

Training data: (7840, 16)
Testing data: (1960, 16)


In [5]:
from sklearn.linear_model import LinearRegression

reduced_model = LinearRegression()

reduced_model.fit(X_train_reduced, y_train_reduced)

print("Reduced Linear Regression trained successfully!")

Reduced Linear Regression trained successfully!


In [6]:
from sklearn.metrics import mean_absolute_error, r2_score

reduced_train_pred = reduced_model.predict(X_train_reduced)
reduced_test_pred = reduced_model.predict(X_test_reduced)

reduced_train_mae = mean_absolute_error(y_train_reduced, reduced_train_pred)
reduced_test_mae = mean_absolute_error(y_test_reduced, reduced_test_pred)

reduced_train_r2 = r2_score(y_train_reduced, reduced_train_pred)
reduced_test_r2 = r2_score(y_test_reduced, reduced_test_pred)

print("Model WITHOUT Sub-Category")
print("--------------------------")
print("Train MAE:", reduced_train_mae)
print("Test MAE:", reduced_test_mae)
print("Train R²:", reduced_train_r2)
print("Test R²:", reduced_test_r2)

Model WITHOUT Sub-Category
--------------------------
Train MAE: 228.80951471018844
Test MAE: 274.619607119606
Train R²: 0.05375050472100196
Test R²: 0.04116153229444974


In [9]:
comparison_day14 = pd.DataFrame({
    'Model': ['Original Linear Regression', 'Without Sub-Category'],
    'Test MAE': [238.769111, 274.619607],
    'Test R²': [0.178595, 0.041162]
})

comparison_day14

,Model,Test MAE,Test R²
0,Original Linear Regression,238.769111,0.178595
1,Without Sub-Category,274.619607,0.041162


In [10]:
mae_change = 274.619607 - 238.769111
r2_change = 0.178595 - 0.041162

print("Increase in Test MAE:", mae_change)
print("Decrease in Test R²:", r2_change)

Increase in Test MAE: 35.850495999999964
Decrease in Test R²: 0.137433


## Sprint 2 Reflection

The model performance decreased after removing the Sub-Category feature.

- Test MAE increased by 35.85
- Test R² decreased by 0.1374

This shows that the model depended significantly on the Sub-Category feature. When an important feature was removed, the system adapted but its predictive performance became weaker.

This experiment helped me understand feature dependency and the importance of building ML systems that can handle changing real-world constraints.